satpy/readers/tropomi_l2.py

```diff
@@ -206,9 +206,25 @@ def get_dataset(self, ds_id, ds_info):
         file_key = ds_info.get('file_key', ds_id.name)
         data = self[file_key]
         data.attrs = self.get_metadata(data, ds_info)
-        fill = data.attrs.pop('_FillValue')
+        fill_value = data.attrs.get('_FillValue', np.float32(np.nan))
         data = data.squeeze()
-        data = data.where(data != fill)
+
+        # preserve integer data types if possible
+        if np.issubdtype(data.dtype, np.integer):
+            new_fill = fill_value
+        else:
+            if np.issubdtype(data.dtype, np.datetime64):
+                new_fill = np.datetime64('NaT')
+            else:
+                new_fill = np.float32(np.nan)
+            data.attrs.pop('_FillValue', None)
+        good_mask = data != fill_value
+
+        scale_factor = data.attrs.get('scale_factor')
+        if scale_factor is not None:
+            data = data * scale_factor
```

Comment:
Is there ever an `add_offset` in these files?

Reply:
Yes, added.